File: packages/runtime-web/src/lifecycle.ts
export type Awaitable<T> = T | PromiseLike<T>

/** The four host-controlled hooks exported by an Extension remote. */
export interface ExtensionModule {
  initialize?(): Awaitable<void>
  activate?(): Awaitable<void>
  deactivate?(): Awaitable<void>
  dispose?(): Awaitable<void>
}

export type LoadedExtensionModule = ExtensionModule | { readonly default: ExtensionModule }
export type ExtensionLoader = () => Awaitable<LoadedExtensionModule>

export const ExtensionState = {
  IDLE: 'IDLE',
  LOADING: 'LOADING',
  LOADED: 'LOADED',
  INITIALIZING: 'INITIALIZING',
  READY: 'READY',
  ACTIVATING: 'ACTIVATING',
  ACTIVE: 'ACTIVE',
  DEACTIVATING: 'DEACTIVATING',
  DISPOSING: 'DISPOSING',
  DISPOSED: 'DISPOSED',
  ERROR: 'ERROR',
} as const

export type ExtensionState = (typeof ExtensionState)[keyof typeof ExtensionState]

export interface ExtensionRuntimeState {
  readonly status: ExtensionState
  readonly error: Error | null
}

export class LifecycleTransitionError extends Error {
  constructor(operation: 'enable' | 'disable', state: ExtensionState) {
    super(`Cannot ${operation} an Extension from lifecycle state ${state}`)
    this.name = 'LifecycleTransitionError'
  }
}

/**
 * Owns only volatile runtime state. Persisting an enabled peer binding remains
 * the host adapter's responsibility after enable() resolves successfully.
 */
export class ExtensionLifecycleController {
  readonly #loader: ExtensionLoader
  #runtimeState: ExtensionRuntimeState = Object.freeze({
    status: ExtensionState.IDLE,
    error: null,
  })
  #module: ExtensionModule | null = null

  constructor(loader: ExtensionLoader) {
    this.#loader = loader
  }

  get state(): ExtensionRuntimeState {
    return this.#runtimeState
  }

  get module(): ExtensionModule | null {
    return this.#module
  }

  async enable(): Promise<void> {
    if (
      this.state.status !== ExtensionState.IDLE &&
      this.state.status !== ExtensionState.DISPOSED
    ) {
      throw new LifecycleTransitionError('enable', this.state.status)
    }

    let initializationStarted = false
    let activationStarted = false

    try {
      this.#transition(ExtensionState.LOADING)
      this.#module = unwrapExtensionModule(await this.#loader())
      this.#transition(ExtensionState.LOADED)

      initializationStarted = true
      this.#transition(ExtensionState.INITIALIZING)
      await this.#module.initialize?.()
      this.#transition(ExtensionState.READY)

      activationStarted = true
      this.#transition(ExtensionState.ACTIVATING)
      await this.#module.activate?.()
      this.#transition(ExtensionState.ACTIVE)
    } catch (cause) {
      const primaryError = asError(cause)
      const cleanupErrors = await this.#compensateFailedEnable(
        initializationStarted,
        activationStarted,
      )
      const failure = combineErrors('Extension enable failed', primaryError, cleanupErrors)
      this.#module = null
      this.#transition(ExtensionState.ERROR, failure)
      throw failure
    }
  }

  async disable(): Promise<void> {
    if (this.state.status !== ExtensionState.ACTIVE || this.#module === null) {
      throw new LifecycleTransitionError('disable', this.state.status)
    }

    const failures: Error[] = []

    this.#transition(ExtensionState.DEACTIVATING)
    try {
      await this.#module.deactivate?.()
    } catch (cause) {
      failures.push(asError(cause))
    }

    this.#transition(ExtensionState.DISPOSING)
    try {
      await this.#module.dispose?.()
    } catch (cause) {
      failures.push(asError(cause))
    }

    if (failures.length > 0) {
      const [primaryError, ...cleanupErrors] = failures
      const failure = combineErrors('Extension disable failed', primaryError, cleanupErrors)
      this.#transition(ExtensionState.ERROR, failure)
      throw failure
    }

    this.#module = null
    this.#transition(ExtensionState.DISPOSED)
  }

  async #compensateFailedEnable(
    initializationStarted: boolean,
    activationStarted: boolean,
  ): Promise<Error[]> {
    if (this.#module === null) {
      return []
    }

    const failures: Error[] = []

    if (activationStarted) {
      try {
        await this.#module.deactivate?.()
      } catch (cause) {
        failures.push(asError(cause))
      }
    }

    if (initializationStarted) {
      try {
        await this.#module.dispose?.()
      } catch (cause) {
        failures.push(asError(cause))
      }
    }

    return failures
  }

  #transition(status: ExtensionState, error: Error | null = null): void {
    this.#runtimeState = Object.freeze({ status, error })
  }
}

function unwrapExtensionModule(loaded: LoadedExtensionModule): ExtensionModule {
  if (typeof loaded !== 'object' || loaded === null) {
    throw new TypeError('Extension loader returned no module')
  }

  const module = 'default' in loaded ? loaded.default : loaded
  if (typeof module !== 'object' || module === null) {
    throw new TypeError('Extension loader returned an invalid module')
  }

  for (const hook of ['initialize', 'activate', 'deactivate', 'dispose'] as const) {
    if (module[hook] !== undefined && typeof module[hook] !== 'function') {
      throw new TypeError(`Extension hook ${hook} must be a function`)
    }
  }

  return module
}

function asError(cause: unknown): Error {
  return cause instanceof Error ? cause : new Error(String(cause))
}

function combineErrors(
  message: string,
  primaryError: Error,
  cleanupErrors: readonly Error[],
): Error {
  if (cleanupErrors.length === 0) {
    return primaryError
  }

  return new AggregateError([primaryError, ...cleanupErrors], `${message}: ${primaryError.message}`)
}

File: src/inkcre_extension_registry/contracts/lifecycle.py
from __future__ import annotations

import inspect
from collections.abc import Awaitable, Callable
from contextlib import suppress
from enum import StrEnum
from typing import Protocol


class ExtensionState(StrEnum):
    DISCOVERED = "DISCOVERED"
    LOADING = "LOADING"
    LOADED = "LOADED"
    INITIALIZING = "INITIALIZING"
    READY = "READY"
    ACTIVATING = "ACTIVATING"
    ACTIVE = "ACTIVE"
    DEACTIVATING = "DEACTIVATING"
    DISPOSING = "DISPOSING"
    UNLOADED = "UNLOADED"
    ERROR = "ERROR"


class ExtensionHooks(Protocol):
    def initialize(self) -> Awaitable[None] | None: ...

    def activate(self) -> Awaitable[None] | None: ...

    def deactivate(self) -> Awaitable[None] | None: ...

    def dispose(self) -> Awaitable[None] | None: ...


async def _maybe_await(value: Awaitable[None] | None) -> None:
    if inspect.isawaitable(value):
        await value


class ExtensionLifecycle[HooksT: ExtensionHooks]:
    """Small host-independent executor for the stable Extension hook contract."""

    def __init__(self, loader: Callable[[], Awaitable[HooksT] | HooksT]) -> None:
        self._loader = loader
        self.hooks: HooksT | None = None
        self.state = ExtensionState.DISCOVERED
        self.error: Exception | None = None

    async def enable(self) -> None:
        if self.state not in {ExtensionState.DISCOVERED, ExtensionState.UNLOADED}:
            raise RuntimeError(f"cannot enable extension from {self.state}")
        activation_started = False
        try:
            self.state = ExtensionState.LOADING
            loaded = self._loader()
            hooks = await loaded if inspect.isawaitable(loaded) else loaded
            self.hooks = hooks
            self.state = ExtensionState.LOADED
            self.state = ExtensionState.INITIALIZING
            await _maybe_await(hooks.initialize())
            self.state = ExtensionState.READY
            self.state = ExtensionState.ACTIVATING
            activation_started = True
            await _maybe_await(hooks.activate())
            self.state = ExtensionState.ACTIVE
            self.error = None
        except Exception as error:
            self.error = error
            if self.hooks is not None:
                if activation_started:
                    with suppress(Exception):
                        await _maybe_await(self.hooks.deactivate())
                with suppress(Exception):
                    await _maybe_await(self.hooks.dispose())
            self.hooks = None
            self.state = ExtensionState.ERROR
            raise

    async def disable(self) -> None:
        if self.state != ExtensionState.ACTIVE or self.hooks is None:
            raise RuntimeError(f"cannot disable extension from {self.state}")
        failures: list[Exception] = []
        hooks = self.hooks
        self.state = ExtensionState.DEACTIVATING
        try:
            await _maybe_await(hooks.deactivate())
        except Exception as error:
            failures.append(error)
        self.state = ExtensionState.DISPOSING
        try:
            await _maybe_await(hooks.dispose())
        except Exception as error:
            failures.append(error)
        if not failures:
            self.hooks = None
            self.state = ExtensionState.UNLOADED
            self.error = None
            return
        self.error = failures[0]
        self.state = ExtensionState.ERROR
        if len(failures) == 1:
            raise failures[0]
        raise ExceptionGroup("extension disable failed", failures)
